nanocoat/lib/scritchui/scritchPencilRasterChunky.c

```diff
@@ -84,22 +84,111 @@ sjme_errorCode sjme_scritchpen_core_copyArea(
 	sjme_attrInValue sjme_jint anchor)
 {
 	sjme_errorCode error;
+	sjme_jint tmpSx, tmpSy, scanlineSize;
+	sjme_cpointer copiedArea;
 	
 	if (g == NULL)
 		return SJME_ERROR_NULL_ARGUMENTS;
+
+	/* The source rectangle must always be in bounds. */
+	if (sx < 0 || sx < 0 || sx + w > g->width || sy + h > g->height)
+		return SJME_ERROR_INDEX_OUT_OF_BOUNDS;
+
+	/* Drawing nothing? */
+	if (w <= 0 || h <= 0)
+		return SJME_ERROR_NONE;
 	
 	/* Need to lock? */
 	if (sjme_error_is(error = sjme_scritchpen_core_lock(g)))
 		return sjme_error_default(error);
 	
-	sjme_todo("Impl?");
-	return sjme_error_notImplemented(0);
+	error = SJME_ERROR_NONE;
+
+	/* Those are needed to translate back before moving to the dest area. */
+	tmpSx = -sx;
+	tmpSy = -sy;
+
+	/* Translate base coordinates. */
+	sjme_scritchpen_coreUtil_applyTranslate(g, &sx, &sy);
+
+	/**
+	 * A neat property here is that we don't need to worry about format
+	 * conversion, as the data is being copied between regions of the same
+	 * image. We can get away with just figuring out the native image format,
+	 * and then go straight to copying from there.
+	 */
+	scanlineSize = (w * g->bitsPerPixel) / 8;
```

Comment:
`scanlineSize = g->scanLenBytes`.

Reply:
Actually no, nevermind. You should use `g->util->pfScanBytes(g, g->pixelFormat, w, -1, &scanlineSize, NULL)`.